cpp/examples/arrow/exec_plan_examples.cc

```diff
@@ -0,0 +1,1122 @@
+  }
+  return *arrow::Table::FromRecordBatches(schema, std::move(batches));
+}
+
+std::shared_ptr<arrow::Table> CreateTable() {
+    auto schema =
+        arrow::schema({arrow::field("a", arrow::int64()),
+        arrow::field("b", arrow::int64()),
+        arrow::field("c", arrow::int64())});
+    std::shared_ptr<arrow::Array> array_a;
+    std::shared_ptr<arrow::Array> array_b;
+    std::shared_ptr<arrow::Array> array_c;
+    arrow::NumericBuilder<arrow::Int64Type> builder;
+    ABORT_ON_FAILURE(builder.AppendValues({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}));
+    ABORT_ON_FAILURE(builder.Finish(&array_a));
+    builder.Reset();
+    ABORT_ON_FAILURE(builder.AppendValues({9, 8, 7, 6, 5, 4, 3, 2, 1, 0}));
+    ABORT_ON_FAILURE(builder.Finish(&array_b));
+    builder.Reset();
+    ABORT_ON_FAILURE(builder.AppendValues({1, 2, 1, 2, 1, 2, 1, 2, 1, 2}));
+    ABORT_ON_FAILURE(builder.Finish(&array_c));
+    return arrow::Table::Make(schema, {array_a, array_b, array_c});
+}
+
+std::shared_ptr<arrow::dataset::Dataset> CreateDataset() {
+    return std::make_shared<arrow::dataset::InMemoryDataset>(
+        GetTableFromJSON(arrow::schema({arrow::field("a", arrow::int32()),
+                                        arrow::field("b", arrow::boolean())}),
+                        {
+                            R"([{"a": 1,    "b": null},
+                                {"a": 2,    "b": true}])",
+                            R"([{"a": null, "b": true},
+                                {"a": 3,    "b": false}])",
+                            R"([{"a": null, "b": true},
+                                {"a": 4,    "b": false}])",
+                            R"([{"a": 5,    "b": null},
+                                {"a": 6,    "b": false},
+                                {"a": 7,    "b": false},
+                                {"a": 8,    "b": true}])",
+                        }));
+}
+
+std::shared_ptr<arrow::Table> CreateSimpleTable() {
+    auto schema = arrow::schema(
+        {arrow::field("a", arrow::int32()), arrow::field("b", arrow::boolean())});
+    std::shared_ptr<arrow::Array> array_a;
+    std::shared_ptr<arrow::Array> array_b;
+    arrow::NumericBuilder<arrow::Int32Type> builder;
+    arrow::BooleanBuilder b_builder;
+    ABORT_ON_FAILURE(builder.AppendValues({1, 2, 3, 4, 5, 6, 7}));
+    ABORT_ON_FAILURE(builder.Finish(&array_a));
+    builder.Reset();
+
+    std::vector<bool> bool_vec{false, true, false, true, false, true, false};
+    ABORT_ON_FAILURE(b_builder.AppendValues(bool_vec));
+    ABORT_ON_FAILURE(builder.Finish(&array_b));
+    builder.Reset();
+    return arrow::Table::Make(schema, {array_a, array_b});
+}
+
+arrow::Status exec_plan_end_to_end_sample() {
+    cp::ExecContext exec_context(arrow::default_memory_pool(),
+                                ::arrow::internal::GetCpuThreadPool());
+
+    // ensure arrow::dataset node factories are in the registry
+    arrow::dataset::internal::Initialize();
```

Comment:
Does this not get called automatically? User code should never have to call this.